sdk/cosmosdb/cosmos/src/client/V2ChangeFeed/ChangeFeedEpkRange.ts

```diff
@@ -0,0 +1,327 @@
+          this.queue.enqueue(feedRange);
+          this.continuationTokenMap.addFeedRange(feedRange);
+        }
+      }
+    }
+  }
+
+  private continuationTokenRidMatchContainerRid(): boolean {
+    if (this.continuationToken.rid !== this.rId) {
+      return false;
+    }
+    return true;
+  }
+
+  public async fetchAllFeedRanges(): Promise<void> {
+    try {
+      const pkRanges = (
+        await this.partitionKeyRangeCache.onCollectionRoutingMap(this.url)
+      ).getOrderedParitionKeyRanges();
+      for (const pkRange of pkRanges) {
+        const feedRange: ChangeFeedRange = new ChangeFeedRange(
+          pkRange.minInclusive,
+          pkRange.maxExclusive,
+          ""
+        );
+        this.queue.enqueue(feedRange);
+        this.continuationTokenMap.addFeedRange(feedRange);
+      }
+    } catch (err) {
+      throw new Error(err.message);
+    }
+  }
+
+  public async fetchOverLappingFeedRanges(pkRange: PartitionKeyRange): Promise<void> {
+    const queryRange = new QueryRange(pkRange.minInclusive, pkRange.maxExclusive, true, false);
+    try {
+      const overLappingRanges = await this.partitionKeyRangeCache.getOverlappingRanges(
+        this.url,
+        queryRange
+      );
+
+      for (const overLappingRange of overLappingRanges) {
+        const feedRange: ChangeFeedRange = new ChangeFeedRange(
+          overLappingRange.minInclusive,
+          overLappingRange.maxExclusive,
+          ""
+        );
+        this.queue.enqueue(feedRange);
+        this.continuationTokenMap.addFeedRange(feedRange);
+      }
+    } catch (err) {
+      throw new ErrorResponse(err.message);
+    }
+  }
+  /**
+   * Gets a value indicating whether there are potentially additional results that can be retrieved.
+   *
+   * Initially returns true. This value is set based on whether there are any remaining unread epkRanges left in the queue.
+   *
+   * @returns Boolean value representing if whether there are potentially additional results that can be retrieved.
+   */
+  get hasMoreResults(): boolean {
+    return !this.queue.isEmpty();
+  }
+
+  /**
+   * Gets an async iterator which will yield pages of results from Azure Cosmos DB.
+   */
+  public async *getAsyncIterator(): AsyncIterable<ChangeFeedResponseV2<Array<T & Resource>>> {
+    do {
+      const result = await this.fetchNext();
+      if (result.count > 0) {
+        yield result;
+      }
+    } while (this.hasMoreResults);
+  }
+
+  private removeFeedRange(): void {
+    this.queue.dequeue();
+  }
+
+  /**
+   * Read feed and retrieves the next page of results in Azure Cosmos DB.
+   */
+  public async fetchNext(): Promise<ChangeFeedResponseV2<Array<T & Resource>>> {
+    const feedRange = this.queue.peek();
+    if (feedRange) {
+      const result = await this.getFeedResponse(feedRange);
+
+      const shouldRetry: boolean = await this.shouldRetryOnFailure(feedRange, result);
+
+      if (shouldRetry) {
+        this.continuationTokenMap.deleteFeedRange(feedRange);
```

Comment:
Feed ranges should not be removed, otherwise how does the user keep checking in the future if there are new changes?